runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/DistributionData.java

```diff
@@ -17,48 +17,283 @@
+  public void computePercentiles() {
+    if (!sketch.isPresent() || sketch.get().getN() == 0) {
+      // if the sketch is not present or is empty, do not compute the percentile
+      return;
+    }
+    this.percentileResults =
+        percentileExtractors.entrySet().stream()
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    Map.Entry::getKey,
+                    kv -> {
+                      final Function<DoublesSketch, Double> metricExtractor = kv.getValue();
+                      return metricExtractor.apply(sketch.get());
+                    }));
+  }
+
+  /** Generates DistributionResult from DistributionData. */
   public DistributionResult extractResult() {
-    return DistributionResult.create(sum(), count(), min(), max());
+    return DistributionResult.create(sum(), count(), min(), max(), percentiles());
+  }
+
+  @Override
+  public boolean equals(Object object) {
+    if (object instanceof DistributionData) {
+      DistributionData other = (DistributionData) object;
+      return Objects.equals(max, other.max())
+          && Objects.equals(min, other.min())
+          && Objects.equals(count, other.count())
+          && Objects.equals(sum, other.sum())
+          && Objects.equals(percentiles(), other.percentiles());
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(min, max, sum, count, percentiles());
+  }
+
+  @Override
+  public String toString() {
+    return "DistributionData{"
+        + "sum="
+        + sum
+        + ", "
+        + "count="
+        + count
+        + ", "
+        + "min="
+        + min
+        + ", "
+        + "max="
+        + max
+        + ", "
+        + "percentiles="
+        + percentileResults
+        + "}";
+  }
+
+  private void writeObject(ObjectOutputStream out) throws IOException {
+    out.defaultWriteObject();
+    if (sketch.isPresent()) {
+      byte[] bytes = sketch.get().toByteArray();
+      out.writeInt(bytes.length);
+      out.write(bytes);
+    }
+  }
+
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  private void readObject(ObjectInputStream in) throws ClassNotFoundException, IOException {
+    in.defaultReadObject();
+    if (!this.percentiles.isEmpty()) {
+      int len = in.readInt();
+      byte[] bytes = new byte[len];
+      in.read(bytes);
+      this.sketch = Optional.of(UpdateDoublesSketch.heapify(Memory.wrap(bytes)));
+      this.percentileExtractors = createPercentileExtractors(this.percentiles);
+    } else {
+      this.sketch = Optional.empty();
+      this.percentileExtractors = ImmutableMap.of();
+    }
+  }
+
+  /** Creates a Map of percentile names and its percentile extractor. */
+  private Map<Double, Function<DoublesSketch, Double>> createPercentileExtractors(
+      Set<Double> customPercentiles) {
+    return customPercentiles.stream()
+        .filter(percentile -> percentile >= 0.0 && percentile <= 100.0)
```

Comment:
Let's do a checkArguments() when the users pass in the customized percentiles, instead of doing the filtering here. The error checking should be explicit and we should throw if it's not right.